apis/v1alpha1/tcproute_types.go

```diff
@@ -19,38 +19,91 @@ import (
 
-	Spec   TcpRouteSpec   `json:"spec,omitempty" protobuf:"bytes,3,opt,name=spec"`
-	Status TcpRouteStatus `json:"status,omitempty" protobuf:"bytes,4,opt,name=status"`
+	Spec   TCPRouteSpec   `json:"spec,omitempty" protobuf:"bytes,3,opt,name=spec"`
+	Status TCPRouteStatus `json:"status,omitempty" protobuf:"bytes,4,opt,name=status"`
+}
+
+// TCPRouteRule is the configuration for a given rule.
+type TCPRouteRule struct {
+	// Match defines which connections match this rule.
+	//
+	// +optional
+	Match *TCPRouteMatch `json:"match" protobuf:"bytes,1,opt,name=match"`
+	// Action defines what happens to the connection.
+	//
+	// +optional
+	Action *TCPRouteAction `json:"action" protobuf:"bytes,2,opt,name=action"`
+}
+
+// TCPRouteAction is the action for a given rule.
+type TCPRouteAction struct {
+	// ForwardTo sends requests to the referenced object.  The
+	// resource may be "services" (omit or use the empty string for the
+	// group), or an implementation may support other resources (for
+	// example, resource "myroutetargets" in group "networking.acme.io").
+	// Omitting or specifying the empty string for both the resource and
+	// group indicates that the resource is "services".  If the referent
+	// cannot be found, the "InvalidRoutes" status condition on any Gateway
+	// that includes the TCPRoute will be true.
+	ForwardTo *ForwardToTarget `json:"forwardTo" protobuf:"bytes,1,opt,name=forwardTo"`
+
+	// ExtensionRef is an optional, implementation-specific extension to the
+	// "action" behavior.  The resource may be "configmaps" (use the empty
+	// string for the group) or an implementation-defined resource (for
+	// example, resource "myrouteactions" in group "networking.acme.io").
+	// Omitting or specifying the empty string for both the resource and
+	// group indicates that the resource is "configmaps".  If the referent
+	// cannot be found, the "InvalidRoutes" status condition on any Gateway
+	// that includes the TCPRoute will be true.
+	//
+	// Support: custom
+	//
+	// +optional
+	ExtensionRef *RouteActionExtensionObjectReference `json:"extensionRef" protobuf:"bytes,2,opt,name=extensionRef"`
+}
+
+// TCPRouteMatch defines the predicate used to match connections to a
+// given action.
+type TCPRouteMatch struct {
+	// ExtensionRef is an optional, implementation-specific extension to the
+	// "match" behavior.  The resource may be "configmap" (use the empty
+	// string for the group) or an implementation-defined resource (for
+	// example, resource "myroutematchers" in group "networking.acme.io").
+	// Omitting or specifying the empty string for both the resource and
+	// group indicates that the resource is "configmaps".  If the referent
+	// cannot be found, the "InvalidRoutes" status condition on any Gateway
+	// that includes the TCPRoute will be true.
+	//
+	// Support: custom
+	//
+	// +optional
+	ExtensionRef *RouteMatchExtensionObjectReference `json:"extensionRef" protobuf:"bytes,1,opt,name=extensionRef"`
```

Comment:
Do we have any asks from the community for an extension point at this level in this resource?
I'm not arguing against this but want to understand the rationale behind it.

Reply:
I think this would be mostly consistency as a pattern.